src/lsm/table_memory.zig

```diff
@@ -0,0 +1,239 @@
+const math = std.math;
+const assert = std.debug.assert;
+
+const constants = @import("../constants.zig");
+const binary_search = @import("binary_search.zig");
+
+pub fn TableMemoryType(comptime Table: type) type {
+    const Key = Table.Key;
+    const Value = Table.Value;
+    const value_count_max = Table.value_count_max;
+    const compare_keys = Table.compare_keys;
+    const key_from_value = Table.key_from_value;
+
+    return struct {
+        const TableMemory = @This();
+
+        pub const ValueContext = struct {
+            count: usize = 0,
+            sorted: bool = true,
+        };
+
+        const Mutability = union(enum) {
+            mutable,
+            immutable: struct {
+                /// An empty table has nothing to flush
+                flushed: bool = true,
+                snapshot_min: u64 = 0,
+            },
+        };
+
+        values: []Value,
+        value_context: ValueContext,
+        mutability: Mutability,
+        name: []const u8,
+
+        pub fn init(
+            allocator: mem.Allocator,
+            mutability: Mutability,
+            name: []const u8,
+        ) !TableMemory {
+            const values = try allocator.alloc(Value, value_count_max);
+            errdefer allocator.free(values);
+
+            return TableMemory{
+                .values = values,
+                .value_context = .{},
+                .mutability = mutability,
+                .name = name,
+            };
+        }
+
+        pub fn deinit(table: *TableMemory, allocator: mem.Allocator) void {
+            allocator.free(table.values);
+        }
+
+        pub fn reset(table: *TableMemory) void {
+            var mutability: Mutability = switch (table.mutability) {
+                .immutable => .{ .immutable = .{} },
+                .mutable => .mutable,
+            };
+
+            table.* = .{
+                .values = table.values,
+                .value_context = .{},
+                .mutability = mutability,
+                .name = table.name,
+            };
+        }
+
+        pub fn count(table: *const TableMemory) usize {
+            return table.value_context.count;
+        }
+
+        pub fn values_used(table: *const TableMemory) []Value {
+            return table.values[0..table.value_context.count];
+        }
+
+        pub fn put(table: *TableMemory, value: *const Value) void {
+            assert(table.mutability == .mutable);
+            assert(table.value_context.count < value_count_max);
+
+            const put_order = if (table.value_context.count == 0)
+                .lt
+            else
+                compare_keys(
+                    key_from_value(&table.values[table.value_context.count - 1]),
+                    key_from_value(value),
+                );
+            table.values[table.value_context.count] = value.*;
+            table.value_context.count += 1;
+
+            table.value_context.sorted = table.value_context.sorted and put_order != .gt;
```

Comment:
(Refactor to avoid a variable and add an assert.)

```suggestion
            if (table.value_context.sorted) {
                table.value_context.sorted = table.value_context.count == 0 or
                    compare_keys(
                        key_from_value(&table.values[table.value_context.count - 1]),
                        key_from_value(value),
                    ).compare(.lte);
            } else {
            assert(table.value_context.count > 0);
            }
                
            table.values[table.value_context.count] = value.*;
            table.value_context.count += 1;
```